packages/cli/src/lib/snapshot/snapshotReporter.ts

```diff
@@ -57,6 +90,115 @@ export class SnapshotReporter {
     );
   }
 
+  private reportHandlers: SnapshotReporterHandlers = {
+    [SnapshotReportStatus.SUCCESS]: () => {},
+    [SnapshotReportStatus.NO_CHANGES]: () => {},
+    [SnapshotReportStatus.MISSING_VAULT_ENTRIES]: () => {},
+    [SnapshotReportStatus.ERROR]: () => {},
+  };
+
+  public setReportHandler<T extends SnapshotReportStatus>(
+    status: T,
+    handler: SnapshotReporterHandlers[T]
+  ): SnapshotReporter {
+    this.reportHandlers[status] = handler;
+    return this;
+  }
+
+  public async handleReport(): Promise<void> {
+    const reportStatuses = this.getReportStatuses();
+    for (const handler of reportStatuses.fixables) {
+      await this.callAndReset(handler);
+    }
+    if (reportStatuses.errors.length > 0) {
+      for (const handler of reportStatuses.errors) {
+        await this.callAndReset(handler);
+      }
+      return;
+    }
+    if (reportStatuses.successes.length > 0) {
+      for (const handler of reportStatuses.successes) {
+        await this.callAndReset(handler);
+      }
+    }
+  }
+
+  private async callAndReset(handler: SnapshotReportStatus) {
+    await this.reportHandlers[handler].apply(this);
+    this.reportHandlers[handler] = () => {};
+  }
+
+  private getReportStatuses() {
+    const statuses: Record<
+      'fixables' | 'successes' | 'errors',
+      SnapshotReportStatus[]
+    > = {
+      fixables: [],
+      successes: [],
+      errors: [],
+    };
+
+    this.computeMissingVaultEntries();
+    if (this.missingVaultEntriesSet.size > 0) {
+      statuses.fixables.push(SnapshotReportStatus.MISSING_VAULT_ENTRIES);
+    }
+
+    if (!this.hasChangedResources()) {
+      statuses.successes.push(SnapshotReportStatus.NO_CHANGES);
+    }
+
+    if (this.isSuccessReport()) {
+      statuses.successes.push(SnapshotReportStatus.SUCCESS);
+    }
+
+    if (this.resourceInErrorCount > 0) {
+      statuses.errors.push(SnapshotReportStatus.ERROR);
+    }
+
+    return statuses;
+  }
+
+  private computeMissingVaultEntries(): void {
+    for (const [resourceType, resource] of Object.entries(
+      this.report.resourceOperationResults
+    )) {
+      for (const errors of Object.values(resource)) {
```

Comment:
There might be a possibility to first extract all errors with a `flatMap`, and then iterate on that.

Would remove the multiple nesting level and improve readability.

Reply:
I thought about it, but I would not do this for two reasons:
1. We currently use two data from the whole object: the resourceType & the message of the error. One is at the very first depth level, the other at the deepest one. Just extracting the error with a `flatMap` would not be enough, and adding the resources to the 'array' generated by the `flatMap` would create more complexity than it'd resolve.
2. If we go in two steps, it means we also go through the array twice. If possible, this is something I'd like to avoid to keep the processing time as low as possible.

Reply:
To alleviates the issue tho, I extracted the scope of the deepest for in another func